app/src/main/java/org/simple/clinic/registration/phone/RegistrationPhoneScreenController.kt

```diff
@@ -53,33 +60,69 @@ class RegistrationPhoneScreenController @Inject constructor(
     return createEmptyEntry.mergeWith(preFill)
   }
 
-  private fun validateInputAndProceed(events: Observable<UiEvent>): Observable<UiChange> {
-    val phoneNumberTextChanges = events.ofType<RegistrationPhoneNumberTextChanged>().map { it.phoneNumber }
-    val doneClicks = events.ofType<RegistrationPhoneDoneClicked>()
-
-    val proceeds = doneClicks
-        .withLatestFrom(phoneNumberTextChanges)
-        .filter { (_, number) -> numberValidator.isValid(number) }
-        .take(1)
-        .flatMap { (_, number) ->
-          userSession.ongoingRegistrationEntry()
-              .map { it.copy(phoneNumber = number) }
-              .flatMapCompletable { userSession.saveOngoingRegistrationEntry(it) }
-              .andThen(Observable.just({ ui: Ui -> ui.openRegistrationNameEntryScreen() }))
-        }
+  private fun showValidationError(events: Observable<UiEvent>): Observable<UiChange> {
+    val phoneNumberTextChanges = events
+        .ofType<RegistrationPhoneNumberTextChanged>()
+        .map { it.phoneNumber }
 
-    val validations = doneClicks
+    return events
+        .ofType<RegistrationPhoneDoneClicked>()
         .withLatestFrom(phoneNumberTextChanges)
         .map { (_, number) -> numberValidator.isValid(number) }
         .filter { isValidNumber -> isValidNumber.not() }
         .map { { ui: Ui -> ui.showInvalidNumberError() } }
-
-    return validations.mergeWith(proceeds)
   }
 
   private fun hideValidationError(events: Observable<UiEvent>): Observable<UiChange> {
     return events
         .ofType<RegistrationPhoneNumberTextChanged>()
-        .map { { ui: Ui -> ui.hideInvalidNumberError() } }
+        .map { { ui: Ui -> ui.hideAnyError() } }
+  }
+
+  private fun saveOngoingEntryAndProceed(events: Observable<UiEvent>): Observable<UiChange> {
+    val phoneNumberTextChanges = events.ofType<RegistrationPhoneNumberTextChanged>().map { it.phoneNumber }
+    val doneClicks = events.ofType<RegistrationPhoneDoneClicked>()
+
+    return doneClicks
+        .withLatestFrom(phoneNumberTextChanges)
+        .filter { (_, number) -> numberValidator.isValid(number) }
+        .flatMap { (_, number) ->
+          val cachedUserFindResult = userSession.findExistingUser(number)
```

Comment:
> Question: Is your intention to cache it only for the duration of the flatMap operator, or for all clicks of the done button?

Only for this `flatMap` because its result is shared between 4 subscribers.